titus-common/src/main/java/com/netflix/titus/common/framework/simplereconciler/internal/EventDistributor.java

```diff
@@ -0,0 +1,270 @@
+import com.netflix.titus.common.util.ExceptionExt;
+import reactor.core.publisher.FluxSink;
+
+/**
+ * {@link EventDistributor} takes over event notifications from the internal reconciler stream, and distributes them
+ * to registered stream observers. Events notifications are processed by a dedicated thread.
+ */
+class EventDistributor<DATA> {
+
+    private static final String ROOT_METRIC_NAME = "titus.simpleReconciliation.eventDistributor.";
+
+    private final Supplier<List<SimpleReconcilerEvent<DATA>>> snapshotSupplier;
+
+    private final LinkedBlockingQueue<EmitterHolder> sinkQueue = new LinkedBlockingQueue<>();
+    private final LinkedBlockingQueue<SimpleReconcilerEvent<DATA>> eventQueue = new LinkedBlockingQueue<>();
+    private final AtomicInteger eventQueueDepth = new AtomicInteger();
+
+    // This is modified only by the internal thread. No need to sync. We use map here for fast removal.
+    // This collection is observed by Spectator poller, so we have to use ConcurrentMap.
+    private final ConcurrentMap<String, EmitterHolder> activeSinks = new ConcurrentHashMap<>();
+
+    private volatile Thread eventLoopThread;
+    private volatile boolean shutdown;
+
+    // Metrics
+    private final Timer metricLoopExecutionTime;
+    private final Counter metricEmittedEvents;
+
+    EventDistributor(Supplier<List<SimpleReconcilerEvent<DATA>>> snapshotSupplier, Registry registry) {
+        this.snapshotSupplier = snapshotSupplier;
+        this.metricLoopExecutionTime = registry.timer(ROOT_METRIC_NAME + "executionTime");
+        PolledMeter.using(registry).withName(ROOT_METRIC_NAME + "eventQueue").monitorValue(this, self -> self.eventQueueDepth.get());
+        PolledMeter.using(registry).withName(ROOT_METRIC_NAME + "activeSubscribers").monitorSize(activeSinks);
+        this.metricEmittedEvents = registry.counter(ROOT_METRIC_NAME + "emittedEvents");
+    }
+
+    void start() {
+        Preconditions.checkState(!shutdown, "Already shutdown");
+        if (eventLoopThread == null) {
+            this.eventLoopThread = new Thread("event-distributor") {
+                @Override
+                public void run() {
+                    doLoop();
+                }
+            };
+            eventLoopThread.setDaemon(true);
+            eventLoopThread.start();
+        }
+    }
+
+    void stop(long timeoutMs) {
+        this.shutdown = true;
+        if (eventLoopThread != null) {
+            try {
+                eventLoopThread.interrupt();
+                if (timeoutMs > 0) {
+                    try {
+                        eventLoopThread.join(timeoutMs);
+                    } catch (InterruptedException ignore) {
+                    }
+                }
+            } finally {
+                eventLoopThread = null;
+            }
+        }
+    }
+
+    void connectSink(FluxSink<List<SimpleReconcilerEvent<DATA>>> sink) {
+        String id = UUID.randomUUID().toString();
```

Comment:
It would be nice if we had a way to add some caller details here for better operational visibility. Then we'd know which caller stream may be having issues. Though I don't see how to do that without plumbing context down through many methods.

Reply:
Yes, I wanted to do exactly that, but could not find any reasonable way to pass this information. We could make it explicit in the `Flux<List<SimpleReconcilerEvent<DATA>>> changes()` method signature (or have a second method with consumer name parameter).